core/src/main/scala/chisel3/Data.scala

```diff
@@ -916,13 +916,15 @@ object Data {
           }
         case (thiz: Record, that: Record) =>
           if (thiz.elements.size != that.elements.size) {
-            false.B
+            throwException(s"Cannot compare Bundles $thiz and $that: types differ")
           } else {
             thiz.elements
               .zip(that.elements)
               .map {
                 case ((thisName, thisData), (thatName, thatData)) =>
-                  thisName.equals(thatName).asBool && thisData === thatData
+                  if (!thisName.equals(thatName))
+                    throwException(s"Cannot compare Bundles $thiz and $that: found differing field names '$thisName' and '$thatName'")
```

Comment:
i dont think we shoudl do it this way since the order does not have to be the same. I'd suggest running down the LHS into a map then looking up the RHS by the keys, similar to https://github.com/chipsalliance/chisel3/blob/master/core/src/main/scala/chisel3/internal/BiConnect.scala#L231

Reply:
Agreed, see my comment here as well https://github.com/chipsalliance/chisel3/pull/2669#discussion_r944038103